Nov 19
下面是找出来的几个候选  
pdfminer3k  
pdfminer.six  
pypdf  
pypdfocr 


In [2]:
# file paths
phone_numbers = ["./data/phone_numbers/070219_KV_BR_Telefonverzei.pdf",
                "./data/phone_numbers/52_pdf_wegweiser_schlaganfall.pdf",
                 "./data/phone_numbers/FL_SYB_BetriebsaerztlicherDienst_ID8414.pdf",
                 "./data/phone_numbers/Infoblatt_154.pdf"
                ]
scans = ["./data/scans/double_ocr.pdf",
         "./data/scans/single_ocr.pdf"]



### pdfminer.six
```
pip install pdfminer.six
```

In [7]:
from pdfminer.pdfinterp import PDFResourceManager, PDFPageInterpreter
from pdfminer.converter import HTMLConverter,TextConverter,XMLConverter
from pdfminer.layout import LAParams
from pdfminer.pdfpage import PDFPage
import io

def convert(case,fname, pages=None):
    if not pages: pagenums = set();
    else:         pagenums = set(pages);      
    manager = PDFResourceManager() 
    codec = 'utf-8'
    caching = True

    if case == 'text' :
        output = io.StringIO()
        converter = TextConverter(manager, output, codec=codec, laparams=LAParams())     
    if case == 'HTML' :
        output = io.BytesIO()
        converter = HTMLConverter(manager, output, codec=codec, laparams=LAParams())

    interpreter = PDFPageInterpreter(manager, converter)   
    infile = open(fname, 'rb')

    for page in PDFPage.get_pages(infile, pagenums,caching=caching, check_extractable=True):
        interpreter.process_page(page)

    convertedPDF = output.getvalue()  

    infile.close(); converter.close(); output.close()
    return convertedPDF

#//////////// main ///////////////////////
filePDF  = phone_numbers[0]     # input
fileHTML = 'myHTML.html'   # output
fileTXT  = 'myTXT.txt'     # output

case = "text"

if case == 'HTML' :
    convertedPDF = convert('HTML', filePDF)
    fileConverted = open(fileHTML, "wb")
if case == 'text' :
    convertedPDF = convert('text', filePDF)
    fileConverted = open(fileTXT, "w", encoding="utf-8")

print(convertedPDF)
fileConverted.write(convertedPDF)
fileConverted.close()
#print(convertedPDF) 



UniversitätsKlinikum Heidelberg

TelefonBuch
So können Sie uns erreichen!
Klinik für Kinder- und Jugendmedizin
· Notfallambulanzen
· Ambulanzen
· Stationäre Aufnahme

Stand Januar 2007
www.klinikum.uni-heidelberg.de

Abteilung Kinderheilkunde I
Schwerpunkt Allgemeine Pädiatrie · Stoffwechsel · 
XXXXXXXX

Inhaltsverzeichnis
Telefonbuch Universitätsklinikum Heidelberg

Telefonzentrale und Notfallambulanzen 

Anästhesiologische Klinik 

Augenklinik 

Chirurgische Klinik 
Klinik für Allgemein-, Viszeral- , Unfall-  und Kinderchirurgie
Klinik für Herzchirurgie
Klinik für Gefäßchirurgie
Klinik für Urologie
 
Frauenklinik 
Allgemeine Frauenheilkunde und Geburtshilfe
Gynäkologische Endokrinologie und Fertilitätsstörungen

Klinik für Hals-, Nasen-, Ohrenheilkunde 

Hautklinik 
 
Heidelberger Ionenstrahl-Therapiezentrum (HIT) 

  42

Klinik für Kinder- und Jugendmedizin 
Kinderheilkunde I
Allgemeine Pädiatrie, Stoffwechsel-Gastroenterologie-Nephrologie
Kinderheilkunde II 
Kinderkardiologie
Kind

### PDF TO Text
For Debian, Ubuntu, and friends
```
sudo apt install build-essential libpoppler-cpp-dev pkg-config python3-dev
```
For Fedora, Red Hat, and friends
```
sudo yum install gcc-c++ pkgconfig poppler-cpp-devel python3-devel
```
For Mac OS
```
brew install pkg-config poppler
```
then run
```
pip install pdftotext
```

In [14]:
import pdftotext

'''
 # Load your PDF
with open(phone_numbers[1], "rb") as f:
    pdf = pdftotext.PDF(f)

 # Save all text to a txt file.
with open('test.txt', 'w') as f:
    f.write("\n\n".join(pdf))
'''

def get_text_pdftotext(pdf):
    with open(pdf, "rb") as f:
        pdf = pdftotext.PDF(f)
        return ("\n\n".join(pdf))
        #for page in pdf:
            #print(page)
print(get_text_pdftotext(phone_numbers[2]))


Betriebsärztlicher Dienst



Untersuchungen auf Wunsch des Beschäftigten                      Kontakt
Nach Arbeitsschutzgesetz hat jeder Arbeitgeber den Beschäf-
tigten eine arbeitsmedizinische Untersuchung zu ermöglichen,     Die Terminvergabe erfolgt
sofern der Mitarbeiter einen möglichen Zusammenhang              Montag – Donnerstag 8:00 – 16:00 Uhr
zwischen seiner Tätigkeit und seiner Erkrankung vermutet.        Freitag                8:00 – 15:00 Uhr
                                                                                                                                                                                                                                                   UniversitätsKlinikum Heidelberg
                                                                 zentral über die Telefonnummern
Verpflichtende Untersuchungen                                    06221 – 56 8966
Wer Tätigkeiten mit einer erhöhten Unfallgefährdung ausübt
oder solche, die mit dem Auf

## phone number format
### phone_numbers[0]
06221 / 56 8781  
06221 / 56 38099  
### phone_numbers[1]
0 62 21 – 56 78 42  
0 72 61 – 66-0  
0 62 21 - 319 – 0  
0 72 61 – 19 29-2  
### phone_numbers[2]
06221 – 56 8966
### phone_numbers[3]
19292  
110  
56-7272  

### Extract Email Address

In [59]:
import re
re_email = re.compile("\w+@[\w-]+[\.\w+]+")

for file in phone_numbers:
    text = get_text_pdftotext(file)
    emails = re_email.findall(text)
    for single_email in emails:
        print(single_email)


carbotti@sportkreis.heidelberg.de
info@aktivija-ev.de
klhauer@web.de
hannelore@web.de
selbsthilfe@paritaet-hd.de
SHV@gmx.de
info@bagcbf.de
info@schlaganfall-hilfe.de
Heidelberg@arbeitsagentur.de
info@aktivija-ev.de
hospiz@ekihd.de
heidelberg@compuserve.de
poster@bethanien-heidelberg.de
hd@skm-heidelberg.de
info@aphasiker.de
caritas@caritas-heidelberg.de
heidelberg@drv-bw.de
mannheim@drv-bw.de
info@drv-bund.de
kontakt@paritaet-hd.de
info@dve.info
info@physio-verband.de
info@drk-rn-heidelberg.de
diakonie@ekihd.de
sinsheim@grn.de
Louise@st.josefskrankenhaus.de
lebenskreis@aol.com
Frien@kvbawue.de
info@kliniken-schmieder.de
info@srh.de
poststelle@vhd.lvs.bwl.de
info@malteser-heidelberg.de
klhauer@web.de
szaltstadt@ekihd.de
szbergheim@t-online.de
reichenbach@drk-rn-heidelberg.de
szkirchheim@psd-ggmbH.de
szneuenheim@psd-ggmbH.de
szpfaffengrund@ekihd.de
szrohrbach@ekihd.de
szweststadt@heidelberg.de
szwieblingen@web.de
ziegelhausen@t-online.de
info@sozialstation-hds.de
vitus@kath-hd.de
diakoni